File: spec/javascript/web_socket/client_spec.js
JS.ENV.WebSocket = JS.ENV.WebSocket || {}

JS.ENV.WebSocketSteps = JS.ENV.IntegrationSteps = JS.Test.asyncSteps({
  server: function(port, callback) {
    this._adapter = new Faye.NodeAdapter({mount: "/bayeux", timeout: 25})
    this._adapter.listen(port)
    this._port = port
    setTimeout(callback, 100)
  },
  
  stop: function(callback) {
    this._adapter.stop()
    setTimeout(callback, 100)
  },
  
  open_socket: function(url, callback) {
    var done = false,
        self = this,
        
        resume = function(open) {
                   if (done) return
                   done = true
                   self._open = open
                   callback()
                 }
    
    this._ws = new Faye.WebSocket.Client(url)
    
    this._ws.onopen  = function() { resume(true)  }
    this._ws.onclose = function() { resume(false) }
  },
  
  close_socket: function(callback) {
    var self = this
    this._ws.onclose = function() {
      self._open = false
      callback()
    }
    this._ws.close()
  },
  
  check_open: function(callback) {
    this.assert( this._open )
    callback()
  },
  
  check_closed: function(callback) {
    this.assert( !this._open )
    callback()
  },
  
  listen_for_message: function(callback) {
    var self = this
    this._ws.onmessage = function(message) { self._message = message.data }
    callback()
  },
  
  send_subscribe: function(callback) {
    this._ws.send(JSON.stringify({channel: "/meta/subscribe", subscription: "/ws"}))
    setTimeout(callback, 100)
  },
  
  check_subscribe_response: function(callback) {
    this.assertEqual( [{ channel:      "/meta/subscribe",
                        error:        "402:clientId:Missing required parameter",
                        subscription: "/ws",
                        successful:   false
                      }],
                      JSON.parse(this._message) )
    callback()
  },
  
  check_no_response: function(callback) {
    this.assert( !this._message )
    callback()
  }
})

JS.ENV.WebSocket.ClientSpec = JS.Test.describe("WebSocket.Client", function() { with(this) {
  include(WebSocketSteps)
  
  before(function() { this.server(8000) })
  after (function() { this.stop() })
  
  it("can open a connection", function() { with(this) {
    open_socket("ws://localhost:8000/bayeux")
    check_open()
  }})
  
  it("can close the connection", function() { with(this) {
    open_socket("ws://localhost:8000/bayeux")
    close_socket()
    check_closed()
  }})
  
  describe("in the OPEN state", function() { with(this) {
    before(function() { with(this) {
      open_socket("ws://localhost:8000/bayeux")
    }})
    
    it("can send and receive messages", function() { with(this) {
      listen_for_message()
      send_subscribe()
      check_subscribe_response()
    }})
  }})
  
  describe("in the CLOSED state", function() { with(this) {
    before(function() { with(this) {
      open_socket("ws://localhost:8000/bayeux")
      close_socket()
    }})
    
    it("cannot send and receive messages", function() { with(this) {
      listen_for_message()
      send_subscribe()
      check_no_response()
    }})
  }})
}})

File: spec/ruby/web_socket/client_spec.rb
# encoding=utf-8

require "spec_helper"

require "thin"
Thin::Logging.silent = true

WebSocketSteps = EM::RSpec.async_steps do
  def server(port, &callback)
    @adapter = Faye::RackAdapter.new(:mount => "/bayeux", :timeout => 25)
    @adapter.listen(port)
    @port = port
    EM.add_timer(0.1, &callback)
  end
  
  def stop(&callback)
    @adapter.stop
    EM.next_tick(&callback)
  end
  
  def open_socket(url, &callback)
    done = false
    
    resume = lambda do |open|
      unless done
        done = true
        @open = true
        callback.call
      end
    end
    
    @ws = Faye::WebSocket::Client.new(url)
    
    @ws.onopen  = lambda { |e| resume.call(true) }
    @ws.onclose = lambda { |e| resume.call(false) }
  end
  
  def close_socket(&callback)
    @ws.onclose = lambda do |e|
      @open = false
      callback.call
    end
    @ws.close
  end
  
  def check_open(&callback)
    @open.should == true
    callback.call
  end
  
  def check_closed(&callback)
    @open.should == false
    callback.call
  end
  
  def listen_for_message(&callback)
    @ws.onmessage = lambda { |e| @message = e.data }
    callback.call
  end
  
  def send_sunscription(&callback)
    @ws.send(JSON.dump "channel" => "/meta/subscribe", "subscription" => "/ws")
    EM.add_timer(0.1, &callback)
  end
  
  def check_subscribe_response(&callback)
    JSON.parse(@message).should == [{ "channel"      => "/meta/subscribe",
                                      "successful"   => false,
                                      "error"        => "402:clientId:Missing required parameter",
                                      "subscription" => "/ws"
                                   }]
    callback.call
  end
  
  def check_no_response(&callback)
    @message.should == nil
    callback.call
  end
end

describe Faye::WebSocket::Client do
  include WebSocketSteps
  
  before do
    Faye.ensure_reactor_running!
    server 8000
    sync
  end
  
  after { stop }
  
  it "can open a connection" do
    open_socket "ws://localhost:8000/bayeux"
    check_open
  end
  
  it "can close the connection" do
    open_socket "ws://localhost:8000/bayeux"
    close_socket
    check_closed
  end
  
  describe "in the OPEN state" do
    before { open_socket "ws://localhost:8000/bayeux" }
    
    it "can send and receive messages" do
      listen_for_message
      send_sunscription
      check_subscribe_response
    end
  end
  
  describe "in the CLOSED state" do
    before do
      open_socket "ws://localhost:8000/bayeux"
      close_socket
    end
    
    it "cannot send and receive messages" do
      listen_for_message
      send_sunscription
      check_no_response
    end
  end
end

